src/jumps/jumpTracker.ts

```diff
@@ -0,0 +1,265 @@
+import * as vscode from 'vscode';
+
+import { Jump } from './jump';
+import { Position } from './../common/motion/position';
+
+/**
+ * JumpTracker is a handrolled version of vscode's TextEditorState
+ * in relation to the 'workbench.action.navigateBack' command.
+ */
+export class JumpTracker {
+  private _jumps: Jump[] = [];
+  private _currentJumpNumber = 0;
+
+  /**
+   * When receiving vscode.window.onDidChangeActiveTextEditor messages,
+   * don't record the jump if we initiated the command.
+   *
+   * Either the jump was added, or it was traversing jump history
+   * and shouldn't count as a new jump.
+   */
+  public isJumpingFiles = false;
+
+  /**
+   * All recorded jumps, in the order of occurrence.
+   */
+  public get jumps(): Jump[] {
+    return this._jumps;
+  }
+
+  /**
+   * Current position in the list of jumps.
+   * This will be past last index if not traveling through history.
+   */
+  public get currentJumpNumber(): number {
+    return this._currentJumpNumber;
+  }
+
+  /**
+   * Current jump in the list of jumps.
+   */
+  public get currentJump(): Jump {
+    return this._jumps[this._currentJumpNumber] || null;
+  }
+
+  /**
+   * Current jump in the list of jumps.
+   */
+  public get hasJumps(): boolean {
+    return this._jumps.length > 0;
+  }
+
+  /**
+   * Last jump in list of jumps.
+   */
+  public get end(): Jump | null {
+    return this._jumps[this._jumps.length - 1];
+  }
+
+  /**
+   * First jump in list of jumps.
+   */
+  public get start(): Jump | null {
+    return this._jumps[0];
+  }
+
+  /**
+   * Record that a jump occurred.
+   *
+   * If the current position is back in history,
+   * jumps after this position will be removed.
+   *
+   * @param from - File/position jumped from
+   * @param to - File/position jumped to
+   */
+  public recordJump(from: Jump | null, to?: Jump | null) {
+    if (from && to && from.onSameLine(to)) {
+      return;
+    }
+
+    if (from) {
+      this.clearJumpsOnSameLine(from);
+    }
+
+    if (from && !from.onSameLine(to)) {
+      this._jumps.push(from);
+    }
+
+    this._currentJumpNumber = this._jumps.length;
+
+    this.clearOldJumps();
+  }
+
+  /**
+   * Record that a jump occurred from one file to another.
+   * This is likely only needed on a handler for
+   * vscode.window.onDidChangeActiveTextEditor.
+   *
+   * File jumps have extra checks in place, keeping in mind
+   * whether this plugin initiated the jump, whether the new file is
+   * a legitimate file.
+   *
+   * @param from - File/position jumped from
+   * @param to - File/position jumped to
+   */
+  public recordFileJump(from: Jump | null, to: Jump) {
+    if (this.isJumpingFiles) {
```

Comment:
Is there a potential race here, and does it matter?

Say we call `handleFileJump` twice before `recordFileJump` gets triggered once by the `onDidChangeActiveTextEditor`, we might record the second `handleFileJump` as a real jump.

Reply:
It does matter if that race condition would occur. If you remove that if block, then it will record the `from` location as a jump again, which puts it back in the jump stack as the next location `<C-o>` will take you to. If you `<C-o>` again, same thing happens, so it circles between the two files so you can't get to older jumps.

But I haven't been able to reproduce the potential race condition. I tested by jumping to definition for two files in a row. I then manually `<C-o>` twice and `<C-i>` twice. Worked fine every time. I also tried `2<C-o>` and `2<C-i>` and that works fine as well.

I'd suggest it is fine as it is. I don't think there is a perfect solution for this. We could use a counter instead, but counters can be worse in their own ways if counter change events are missed for whatever reason. I also was going to originally just go by filename, seeing if it matches the previous or next in the  jump list, but that doesn't match vanilla Vim and might have unexpected behavior.

Reply:
Alright, I tested it out, and I think the race condition occurring doesn't cause it to fail disastrously. I'm fine with it.